### PR TITLE
Fix/dpcpp headers location and metadata

### LINKED ISSUE #31: Build failure when using dpcpp compiler and sysroot_linux >=2.28
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

Please refer to https://github.com/oleksandr-pavlyk/conda-build-failure-reproducer for the reproducer, and sample output. 

The reproducer derives from https://github.com/IntelPython/dpctl.git

The reproducer sets `sysroot_linux-64 >=2.28` because Intel oneAPI DPC++ compiler requires GLIBC version 2.28 or newer. Quoting from https://www.intel.com/content/www/us/en/developer/articles/system-requirements/intel-oneapi-dpcpp-system-requirements.html:

> Supported minimum versions of distributed libraries that the compilers are built on: GCC - 7.5.0, BINUTILS- 2.30, GLIBC-2.28, G++ - 7.5.0.

The reproducer runs into weird error about incompatible component merge:

```
    ValueError: Incompatible component merge:
      - '*mpich*'
      - 'mpi_mpich_*'
```

but the root of the problem is that `dpcpp_impl_linux-64` pins `sysroot_linux-64` to 2.17.*:

```
dpcpp_impl_linux-64 2024.0.0 hd6dfaa8_49819
-------------------------------------------
file name   : dpcpp_impl_linux-64-2024.0.0-hd6dfaa8_49819.conda
name        : dpcpp_impl_linux-64
version     : 2024.0.0
build       : hd6dfaa8_49819
build number: 49819
size        : 337.4 MB
license     : LicenseRef-Proprietary-Intel-End-User-License-Agreement-for-Developer-Tools
subdir      : linux-64
url         : https://conda.anaconda.org/conda-forge/linux-64/dpcpp_impl_linux-64-2024.0.0-hd6dfaa8_49819.conda
md5         : 61de3d1697746244be94ddb3d90c00ec
timestamp   : 2023-11-17 14:47:28 UTC
dependencies:
  - _openmp_mutex >=4.5
  - dpcpp-cpp-rt 2024.0.0 h59595ed_49819
  - gcc_impl_linux-64
  - gxx_impl_linux-64
  - libgcc-ng >=12
  - libstdcxx-ng >=12
  - libxml2 >=2.11.5,<3.0.0a0
  - libzlib >=1.2.13,<1.3.0a0
  - ocl-icd >=2.3.1,<3.0a0
  - sysroot_linux-64 2.17.*
```

In fact, replacing `sysroot_linux-64 >=2.28` in the reproducer with `sysroot_linux-64 >=2.17` resolves this failure.

Please fix this by releasing `dpcpp` based on `sysroot_linux-64 2.28.*` 

### Installed packages

```shell
Fresh installation of miniconda3 on Ubuntu 20.04.
```

### Environment info

```shell
active environment : None
            shell level : 0
       user config file : /home/opavlyk/.condarc
 populated config files :
          conda version : 23.7.4
    conda-build version : 3.26.1
         python version : 3.9.12.final.0
       virtual packages : __archspec=1=x86_64
                          __glibc=2.31=0
                          __linux=5.15.133.1=0
                          __unix=0=0
       base environment : /home/opavlyk/miniconda3  (writable)
      conda av data dir : /home/opavlyk/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/opavlyk/miniconda3/pkgs
                          /home/opavlyk/.conda/pkgs
       envs directories : /home/opavlyk/miniconda3/envs
                          /home/opavlyk/.conda/envs
               platform : linux-64
             user-agent : conda/23.7.4 requests/2.31.0 CPython/3.9.12 Linux/5.15.133.1-microsoft-standard-WSL2 ubuntu/20.04.6 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
```

### LINKED ISSUE #32: The Intel compiler breaks conda c++ compiler if installed
### Solution to issue cannot be found in the documentation.

- [X] I checked the documentation.

### Issue

The Intel compiler headers are written to `$CONDA_PREFIX/include` and take priority to `x86_64-conda-linux-gnu-c++` compiler headers. For example, this it true for `float.h` header.

The problem is that this header works like an empty header on Linux when `__INTEL_COMPILER` is not set. For Windows it correctly includes system `float.h` header, for Linux it does not.

### Installed packages

```shell
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                  2_kmp_llvm    conda-forge
_sysroot_linux-64_curr_repodata_hack 3                   h69a702a_14    conda-forge
binutils_impl_linux-64    2.40                 hf600244_0    conda-forge
binutils_linux-64         2.40                 hdade7a5_3    conda-forge
bzip2                     1.0.8                hd590300_5    conda-forge
c-ares                    1.27.0               hd590300_0    conda-forge
ca-certificates           2024.2.2             hbcca054_0    conda-forge
cmake                     3.22.3               h5432695_0    conda-forge
dpcpp-cpp-rt              2024.0.0         h59595ed_49819    conda-forge
dpcpp_impl_linux-64       2024.0.0         h034604b_49819    conda-forge
expat                     2.6.2                h59595ed_0    conda-forge
gcc_impl_linux-64         11.4.0               h7aa1c59_5    conda-forge
gcc_linux-64              11.4.0               h0f0c6b6_3    conda-forge
gxx_impl_linux-64         11.4.0               h7aa1c59_5    conda-forge
gxx_linux-64              11.4.0               h2730b16_3    conda-forge
icu                       73.2                 h59595ed_0    conda-forge
intel-cmplr-lib-rt        2024.0.0         hfc55251_49819    conda-forge
intel-cmplr-lic-rt        2024.0.0         ha770c72_49819    conda-forge
intel-opencl-rt           2024.0.0         heb1d2f6_49819    conda-forge
kernel-headers_linux-64   3.10.0              h4a8ded7_14    conda-forge
keyutils                  1.6.1                h166bdaf_0    conda-forge
krb5                      1.21.2               h659d440_0    conda-forge
ld_impl_linux-64          2.40                 h41732ed_0    conda-forge
libcurl                   8.6.0                hca28451_0    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 hd590300_2    conda-forge
libexpat                  2.6.2                h59595ed_0    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-devel_linux-64     11.4.0             h922705a_105    conda-forge
libgcc-ng                 13.2.0               h807b86a_5    conda-forge
libgomp                   13.2.0               h807b86a_5    conda-forge
libhwloc                  2.9.3           default_h554bfaf_1009    conda-forge
libiconv                  1.17                 hd590300_2    conda-forge
libnghttp2                1.58.0               h47da74e_1    conda-forge
libnsl                    2.0.1                hd590300_0    conda-forge
libsanitizer              11.4.0               h4dcbe23_5    conda-forge
libsqlite                 3.45.2               h2797004_0    conda-forge
libssh2                   1.11.0               h0841786_0    conda-forge
libstdcxx-devel_linux-64  11.4.0             h922705a_105    conda-forge
libstdcxx-ng              13.2.0               h7e041cc_5    conda-forge
libuuid                   2.38.1               h0b41bf4_0    conda-forge
libuv                     1.48.0               hd590300_0    conda-forge
libxcrypt                 4.4.36               hd590300_1    conda-forge
libxml2                   2.12.6               h232c23b_0    conda-forge
libzlib                   1.2.13               hd590300_5    conda-forge
llvm-openmp               18.1.1               h4dfa4b3_0    conda-forge
mkl                       2024.0.0         ha957f24_49657    conda-forge
ncurses                   6.4.20240210         h59595ed_0    conda-forge
ninja                     1.10.2               h4bd325d_1    conda-forge
ocl-icd                   2.3.2                hd590300_0    conda-forge
onednn                    3.3              omp_hc847620_3    conda-forge
onednn-devel              2024.0.0            intel_49521    intel
onemkl-sycl-blas          2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-datafitting   2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-dft           2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-lapack        2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-rng           2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-sparse        2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-stats         2024.0.0         ha957f24_49657    conda-forge
onemkl-sycl-vm            2024.0.0         ha957f24_49657    conda-forge
openssl                   3.2.1                hd590300_1    conda-forge
pip                       23.3.2             pyhd8ed1ab_0    conda-forge
python                    3.9.19          h0755675_0_cpython    conda-forge
readline                  8.2                  h8228510_1    conda-forge
rhash                     1.4.3                hd590300_2    conda-forge
setuptools                69.2.0             pyhd8ed1ab_0    conda-forge
sysroot_linux-64          2.17                h4a8ded7_14    conda-forge
tbb                       2021.11.0            h00ab1b0_1    conda-forge
tk                        8.6.13          noxft_h4845f30_101    conda-forge
tzdata                    2024a                h0c530f3_0    conda-forge
wheel                     0.42.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h166bdaf_0    conda-forge
zlib                      1.2.13               hd590300_5    conda-forge
zstd                      1.5.5                hfc55251_0    conda-forge
```

### Environment info
```
shell
active environment : triton
    active env location : /home/jovyan/.conda/envs/triton
            shell level : 4
       user config file : /home/jovyan/.condarc
 populated config files : /home/jovyan/.conda/.condarc
          conda version : 23.11.0
    conda-build version : not installed
         python version : 3.10.13.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=sapphirerapids
                          __conda=23.11.0=0
                          __glibc=2.35=0
                          __linux=5.15.0=0
                          __unix=0=0
       base environment : /home/jovyan/.conda  (writable)
      conda av data dir : /home/jovyan/.conda/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/jovyan/.conda/pkgs
       envs directories : /home/jovyan/.conda/envs
               platform : linux-64
             user-agent : conda/23.11.0 requests/2.31.0 CPython/3.10.13 Linux/5.15.0-94-generic ubuntu/22.04.3 glibc/2.35 solver/libmamba conda-libmamba-solver/23.12.0 libmambapy/1.5.5
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
These changes are target to resolve #31, #32 and update dependencies the way the compiler may be used independently of other compilers. So users typical usage will look like:
```
build:
  - compiler('dpcpp')
  - stdlib('c')
```

instead of
```
build:
  - compiler('c')
  - compiler('dpcpp')
  - stdlib('c')
```

Changes also allow older versions of gcc be used with the dpcpp compiler, that supported by the compiler: https://www.intel.com/content/www/us/en/developer/articles/system-requirements/intel-oneapi-dpcpp-system-requirements.html .

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Fixes #31
Fixes #32

### CODE REVIEW

**Comment by ZzEeKkAa** [comment]
@conda-forge-admin, please rerender

**Comment by ZzEeKkAa** [comment]
@conda-forge-admin, please rerender

**Comment by ZzEeKkAa** [comment]
New dependency list:
```shell
$ conda search local::dpcpp_impl_linux-64=2024.1.0 --info
Loading channels: done
dpcpp_impl_linux-64 2024.1.0 hc1f2bb5_964
-----------------------------------------
file name   : dpcpp_impl_linux-64-2024.1.0-hc1f2bb5_964.conda
name        : dpcpp_impl_linux-64
version     : 2024.1.0
build       : hc1f2bb5_964
build number: 964
size        : 244.4 MB
license     : LicenseRef-Proprietary-Intel-End-User-License-Agreement-for-Developer-Tools
subdir      : linux-64
url         : file:///home/yevhenii/.miniforge3/conda-bld/linux-64/dpcpp_impl_linux-64-2024.1.0-hc1f2bb5_964.conda
md5         : 734dc428076f5b43b12f41b2f7229c74
timestamp   : 2024-06-26 18:49:33 UTC
dependencies:
  - __glibc >=2.17,<3.0.a0
  - _openmp_mutex >=4.5
  - binutils_impl_linux-64 >=2.30
  - dpcpp-cpp-rt 2024.1.0 he02047a_964
  - libgcc-devel_linux-64 >=7.5.0
  - libgcc-ng >=12
  - libstdcxx-devel_linux-64 >=7.5.0
  - libstdcxx-ng >=12
  - libxml2 >=2.12.7,<3.0a0
  - libzlib >=1.3.1,<2.0a0
  - ocl-icd >=2.3.2,<3.0a0
  - sysroot_linux-64 >=2.17

$conda search local::dpcpp_linux-64=2024.1.0 --info
Loading channels: done
dpcpp_linux-64 2024.1.0 h85509e4_964
------------------------------------
file name   : dpcpp_linux-64-2024.1.0-h85509e4_964.conda
name        : dpcpp_linux-64
version     : 2024.1.0
build       : h85509e4_964
build number: 964
size        : 29 KB
license     : LicenseRef-Proprietary-Intel-End-User-License-Agreement-for-Developer-Tools
subdir      : linux-64
url         : file:///home/yevhenii/.miniforge3/conda-bld/linux-64/dpcpp_linux-64-2024.1.0-h85509e4_964.conda
md5         : 2cdeb3c4b35b1dc4da34b6fc32bc1727
timestamp   : 2024-06-26 19:00:17 UTC
dependencies:
  - binutils_impl_linux-64 >=2.30
  - dpcpp_impl_linux-64 2024.1.0 hc1f2bb5_964
  - sysroot_linux-64
```

File: recipe/repack.sh [hunk 1]
```diff
@@ -6,6 +6,15 @@ set -ex
 
 src="${SRC_DIR}/${PKG_NAME}"
 
+# TODO: remove once fixed in the upstream
+if [[ "$PKG_NAME" == "dpcpp_impl_linux-64" ]]; then
+  # Move intel specific headers files from the global include directory to
+  # allow other compilers work in the same environment.
+  find ${src}/include -maxdepth 1 -type f -exec mv {} ${src}/opt/compiler/include \;
```
Comment (napetrov) [inline recipe/repack.sh:13]: why opt in path?  should this be dpcpp name instead to reflect this is dpcpp specific?

**Review by napetrov — APPROVED** [review]

**Pull request merged**